### PR TITLE
feat: [PROD-11431] prevent unsaved scenario parameters

### PR DESCRIPTION
**Pull request by ardvi**
Before creating a PR, please check that:
- [ ] Code is clean
- [ ] Tests are still working (cypress tests and `yarn test`)
- [ ] Changes don't cause new errors or warnings in browser console
- [ ] Documentation is up-to-date
- [ ] Breaking changes are clearly identified with [conventional commits](https://kapeli.com/cheat_sheets/Conventional_Commits.docset/Contents/Resources/Documents/index)

### CODE REVIEW

**Review by esasova — COMMENTED** [review]
Some little remarks but I have three questions: 

1. While testing the feature, I was able to close the window sometimes, is there some edge cases to know ?
2. The user can disconnect without confirming discard, I don't remember if we decided to include it in the ticket or not
3. I'm not sure, we need to confirm it with @csm-thu  but changing the router can be a breaking change

File: cypress/commons/actions/generic/ScenarioParameters.js [hunk 1]
```diff
@@ -70,8 +78,16 @@ function discard() {
...
+  getParametersDiscardAndContinueButton().click();
+}
+
+function cancelDiscardAndContinue() {
+  getParametersCancelDiscardAndContinueButton().click();
+}
+
 function launch(timeoutGetLaunchButton = 180) {
-  const alias = api.forgeAlias('reqRunScenarioAlias');
+  const alias = api.forgeAlias('reqSaveScenarioAlias');
```
Comment (csm-thu) [inline cypress/commons/actions/generic/ScenarioParameters.js:90]: The previous alias name `reqRunScenarioAlias` seems more adapted for this `launch` function

File: src/components/CurrentScenarioSelector/CurrentScenarioSelector.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  /**
+   *  Define the HierarchicalComboBox's state:
+   *  - true : the button is disabled (the tooltip will guide users on how to enable the button)
+   *  - false : the button is enabled
+   */
+  disabled: PropTypes.bool,
+  /**
+   * Tooltip on HierarchicalComboBox
+   */
+  renderInputToolType: PropTypes.string,
```
Comment (csm-thu) [inline src/components/CurrentScenarioSelector/CurrentScenarioSelector.js:84]: Typo: `renderInputToolType` should be `renderInputTooltip`

File: src/components/CurrentScenarioSelector/CurrentScenarioSelector.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+  const handleScenarioChange = useCallback(
+    (event, scenario) => {
+      changeScenario(scenario.id);
+    },
+    [changeScenario]
+  );
+
+  const noScenario = currentScenarioData === null;
+  const scenarioListDisabled = sortedScenarioList.length === 0 || noScenario;
```
Comment (csm-thu) [inline src/components/CurrentScenarioSelector/CurrentScenarioSelector.js:27]: We could probably improve this part by writing instead:
`const noScenario = currentScenarioData === null || sortedScenarioList.length === 0;`

and using it in the props of the returned componet as:
`disabled={disabled || noScenario || isSwitchingScenario}`

File: src/services/twoActionsDialog/twoActionsDialogService.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  openDialog(dialogProps) {
+    return new Promise((resolve) => {
+      dialogProps.resolve = resolve;
+      twoActionsDialogNotifier.send(dialogProps);
+    });
+  }
+}
+
+const TwoActionsDialogGlobal = () => {
+  const suscription = useRef();
```
Comment (csm-thu) [inline src/services/twoActionsDialog/twoActionsDialogService.js:17]: Typo: `suscription` should be `subscription`

File: src/views/Instance/Instance.js [hunk 5]
```diff
@@ -3,35 +3,28 @@
...
+import { CytoViz } from '@cosmotech/ui';
 import { parseError } from '../../utils/ErrorsUtils';
 import { STATUSES } from '../../state/commons/Constants';
 import { AppInsights } from '../../services/AppInsights';
 import { fetchData, processGraphElements } from './data';
 import useStyles from './style';
+import { CurrentScenarioSelector } from '../../components';
 import { useTheme } from '@mui/styles';
 import { useInstance } from './InstanceHook';
+import { useRedirectionToScenario } from '../../hooks/RouterHooks';
```
Comment (csm-thu) [inline src/views/Instance/Instance.js:15]: This hook has been removed from the `InstanceHook` file to be imported here, was it on purpose ?

File: cypress/integration/brewery/DiscardAndContinue.spec.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
+import { ScenarioManager, Scenarios, ScenarioParameters, Login, Workspaces } from '../../commons/actions';
+import { BreweryParameters } from '../../commons/actions/brewery';
+import { stub } from '../../commons/services/stubbing';
+import utils from '../../commons/TestUtils';
+import { routeUtils as route } from '../../commons/utils';
+import { EXTENDED_WORKSPACES_LIST } from '../../fixtures/stubbing/default';
+
+describe('Discard and continue inside same scenario', () => {
```
Comment (csm-thu) [inline cypress/integration/brewery/DiscardAndContinue.spec.js:8]: Did you mean "inside the same **workspace**"?

File: public/locales/fr/translation.json [hunk 7]
```diff
@@ -500,6 +500,14 @@
             "body": "Lancer un scénario peut prendre du temps. Tous les résultats existants pour ce scénario seront écrasés.",
             "checkbox": "Ne pas demander la prochaine fois."
           }
+        },
+        "discardUnsaved": {
+          "title": "Rejeter les paramètres non sauvegardés ?",
+          "body": "Vous avez des paramètres non sauvés, qui seront perdus si vous continuez.",
```
[the remark this answers is not in the document]
Comment (ardvi) [reply public/locales/fr/translation.json]: done

File: public/locales/en/translation.json [hunk 8]
```diff
@@ -499,6 +499,14 @@
             "body": "Running a scenario can take time. Any existing results for this scenario will be overwriten.",
             "checkbox": "Don't ask again"
           }
+        },
+        "discardUnsaved": {
+          "title": "Discard unsaved parameters ?",
```
[the remark this answers is not in the document]
Comment (ardvi) [reply public/locales/en/translation.json]: done

File: public/locales/fr/translation.json [hunk 9]
```diff
@@ -500,6 +500,14 @@
             "body": "Lancer un scénario peut prendre du temps. Tous les résultats existants pour ce scénario seront écrasés.",
             "checkbox": "Ne pas demander la prochaine fois."
           }
+        },
+        "discardUnsaved": {
+          "title": "Rejeter les paramètres non sauvegardés ?",
```
[the remark this answers is not in the document]
Comment (ardvi) [reply public/locales/fr/translation.json]: done

File: src/components/CurrentScenarioSelector/CurrentScenarioSelector.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+  const handleScenarioChange = useCallback(
+    (event, scenario) => {
+      changeScenario(scenario.id);
+    },
+    [changeScenario]
+  );
+
+  const noScenario = currentScenarioData === null;
+  const scenarioListDisabled = sortedScenarioList.lenght === 0 || noScenario;
```
[the remark this answers is not in the document]
Comment (ardvi) [reply src/components/CurrentScenarioSelector/CurrentScenarioSelector.js]: done

File: src/hooks/RouterHooks.js [hunk 11]
```diff
@@ -1,12 +1,13 @@
...
+import { useCallback, useEffect, useRef } from 'react';
+import { useNavigate, useParams, unstable_useBlocker as useBlocker } from 'react-router-dom';
+import { useFindScenarioById, useCurrentScenarioId } from '../state/hooks/ScenarioHooks';
 import { useWorkspaceData } from '../state/hooks/WorkspaceHooks';
-import { STATUSES } from '../state/commons/Constants';
 import { ConfigUtils } from '../utils';
+import { useSortedScenarioList } from './ScenarioListHooks';
+import { TwoActionsDialogService } from '../services/twoActionsDialog/twoActionsDialogService';
 
 export const useRedirectFromInstanceToSenarioView = () => {
```
[the remark this answers is not in the document]
Comment (ardvi) [reply src/hooks/RouterHooks.js]: done

File: src/hooks/RouterHooks.js [hunk 12]
```diff
@@ -29,47 +30,90 @@ export const useRedirectFromInstanceToSenarioView = () => {
...
+  const currentScenarioId = useCurrentScenarioId();
+
   const navigate = useNavigate();
   const handleScenarioChange = useFindScenarioById();
-  const updateCurrentScenario = useUpdateCurrentScenario();
+  const sortedScenarioList = useSortedScenarioList();
+
+  const routerScenarioId = useRouterScenarioId();
 
+  // no routerid and no current scenarioid > Navigate to the first scenario
```
[the remark this answers is not in the document]
Comment (ardvi) [reply src/hooks/RouterHooks.js]: done

File: src/services/twoActionsDialog/twoActionsDialogService.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+const twoActionsDialogNotifier = new Notifier();
+
+class TwoActionsDialogClass {
+  async openDialog(dialogProps) {
+    const result = await new Promise((resolve) => {
+      dialogProps.resolve = resolve;
+      twoActionsDialogNotifier.send(dialogProps);
+    });
+
+    return result;
```
[the remark this answers is not in the document]
Comment (ardvi) [reply src/services/twoActionsDialog/twoActionsDialogService.js]: done

File: cypress/commons/actions/generic/ScenarioParameters.js (hunk 1 above)
Comment (ardvi) [reply cypress/commons/actions/generic/ScenarioParameters.js]: done

File: src/components/CurrentScenarioSelector/CurrentScenarioSelector.js (hunk 2 above)
Comment (ardvi) [reply src/components/CurrentScenarioSelector/CurrentScenarioSelector.js]: done

File: src/components/CurrentScenarioSelector/CurrentScenarioSelector.js (hunk 3 above)
Comment (ardvi) [reply src/components/CurrentScenarioSelector/CurrentScenarioSelector.js]: done

File: src/hooks/RouterHooks.spec.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+      beforeEach(() => {
+        setUp(true);
+      });
+
+      it(`don't display confirmation on window unload`, async () => {
+        const event = new Event('beforeunload', { cancelable: true });
+        const result = window.dispatchEvent(event);
+        expect(result).toBeFalsy();
+      });
+    });
```
[the remark this answers is not in the document]
Comment (ardvi) [reply src/hooks/RouterHooks.spec.js]: done

File: src/services/twoActionsDialog/twoActionsDialogService.js (hunk 4 above)
Comment (ardvi) [reply src/services/twoActionsDialog/twoActionsDialogService.js]: done

File: src/state/dispatchers/scenario/ScenarioDispatcher.js [hunk 15]
```diff
@@ -32,6 +32,10 @@ export const dispatchCreateScenario = (organizationId, workspaceId, scenario) =>
   scenario,
 });
 
+export const dispatchResetNewScenario = () => ({
+  type: SCENARIO_ACTIONS_KEY.RESET_NEW_SCENARIO,
+});
+
```
[the remark this answers is not in the document]
Comment (ardvi) [reply src/state/dispatchers/scenario/ScenarioDispatcher.js]: done

File: src/state/hooks/ScenarioHooks.js [hunk 16]
```diff
@@ -52,6 +52,14 @@ export const useCurrentScenarioState = () => {
   return useSelector((state) => state.scenario.current?.data?.state);
 };
 
+export const useCurrentScenarioId = () => {
+  return useSelector((state) => state?.scenario?.current?.data?.id);
+};
+
+export const useCurrentScenarioStatus = () => {
```
[the remark this answers is not in the document]
Comment (ardvi) [reply src/state/hooks/ScenarioHooks.js]: done

File: src/views/Instance/Instance.js (hunk 5 above)
Comment (ardvi) [reply src/views/Instance/Instance.js]: done

File: cypress/integration/brewery/DiscardAndContinue.spec.js (hunk 6 above)
Comment (ardvi) [reply cypress/integration/brewery/DiscardAndContinue.spec.js]: done

File: src/components/CurrentScenarioSelector/CurrentScenarioSelector.js [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+  disabled: PropTypes.bool,
+  /**
+   * Tooltip on HierarchicalComboBox
+   */
+  renderInputToolTip: PropTypes.string,
+};
+
+CurrentScenarioSelector.defaultProps = {
+  disabled: false,
+  renderInputToolType: '',
```
Comment (csm-thu) [inline src/components/CurrentScenarioSelector/CurrentScenarioSelector.js:92]: The default prop still has its previous name `renderInputToolType`

**Review by csm-thu — APPROVED** [review]
LGTM :+1:

File: public/locales/en/translation.json [hunk 18]
```diff
@@ -374,7 +374,9 @@
           "label": "Copy",
           "copied": "Copied"
         }
-      }
+      },
+      "openScenario": "Scenario can't be openned.",
```
Comment (csm-thu) [inline public/locales/en/translation.json:378]: Typo: `openned` -> `opened`

File: src/state/sagas/scenario/FindScenarioById/FindScenarioByIdData.js [hunk 19]
```diff
@@ -33,6 +34,18 @@ export function* fetchScenarioByIdData(action) {
     data.parametersValues = ApiUtils.formatParametersFromApi(data.parametersValues);
 
     ScenariosUtils.patchScenarioWithCurrentUserPermissions(data, userEmail, userId, scenariosPermissionsMapping);
+
+    if (!data.security.currentUserPermissions.includes(ACL_PERMISSIONS.SCENARIO.READ)) {
+      const err = new Error(t('commoncomponents.banner.openScenario', "Scenario can't be openned"));
```
Comment (csm-thu) [inline src/state/sagas/scenario/FindScenarioById/FindScenarioByIdData.js:39]: Same typo as in the translation file: `openned` -> `opened`

File: src/components/CurrentScenarioSelector/CurrentScenarioSelector.js (hunk 17 above)
Comment (ardvi) [reply src/components/CurrentScenarioSelector/CurrentScenarioSelector.js]: done

File: public/locales/en/translation.json (hunk 18 above)
Comment (ardvi) [reply public/locales/en/translation.json:378]: done

File: src/state/sagas/scenario/FindScenarioById/FindScenarioByIdData.js (hunk 19 above)
Comment (ardvi) [reply src/state/sagas/scenario/FindScenarioById/FindScenarioByIdData.js:39]: done

File: cypress/integration/brewery/DiscardAndContinue.spec.js [hunk 20]
```diff
@@ -0,0 +1,132 @@
+import { ScenarioManager, Scenarios, ScenarioParameters, Login, Workspaces } from '../../commons/actions';
```
Comment (Guismus) [inline cypress/integration/brewery/DiscardAndContinue.spec.js:1]: Maybe need a header here

File: src/services/twoActionsDialog/twoActionsDialogService.js [hunk 21]
```diff
@@ -0,0 +1,54 @@
+import React, { useEffect, useRef, useState } from 'react';
```
Comment (Guismus) [inline src/services/twoActionsDialog/twoActionsDialogService.js:1]: Maybe need a header here

File: src/components/CurrentScenarioSelector/CurrentScenarioSelector.js [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+     {
+       label: 'string',
+       validationStatus:
+       {
+         rejected: 'string',
+         validated: 'string',
+       }
+     }
+   * </pre>
+   */
```
Comment (Guismus) [inline src/components/CurrentScenarioSelector/CurrentScenarioSelector.js:70]: Comment seems a bit overkill (object just after tell type by himself)

File: src/components/CurrentScenarioSelector/CurrentScenarioSelector.js [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+  labels: PropTypes.shape({
+    label: PropTypes.string,
+    validationStatus: PropTypes.shape({
+      rejected: PropTypes.string,
+      validated: PropTypes.string,
+    }),
+  }),
+  /**
```
Comment (Guismus) [inline src/components/CurrentScenarioSelector/CurrentScenarioSelector.js:78]: Comments seems a bit overkill too

File: src/hooks/RouterHooks.js [hunk 24]
```diff
@@ -29,47 +30,96 @@ export const useRedirectFromInstanceToSenarioView = () => {
...
-      if (routerParameters.scenarioId === undefined) {
-        navigate(`${currentScenario.data.id}`, { replace: true });
-      } else if (currentScenario.data.id !== routerParameters.scenarioId) {
-        handleScenarioChange(routerParameters.scenarioId);
-        if (!sortedScenarioList?.find((scenario) => scenario.id === routerParameters.scenarioId))
-          navigate(`${currentScenario.data.id}`);
-      }
+    if (sortedScenarioList.length !== 0 && !routerScenarioId && !currentScenarioId) {
+      navigate(sortedScenarioList[0].id);
     }
```
Comment (Guismus) [inline src/hooks/RouterHooks.js:51]: useless brackets on if statement

File: src/hooks/RouterHooks.js [hunk 25]
```diff
@@ -29,47 +30,96 @@ export const useRedirectFromInstanceToSenarioView = () => {
...
-    if (currentScenario.data.id !== routerParameters.scenarioId) {
-      const scenarioFromUrl = sortedScenarioList.find((el) => el.id === routerParameters.scenarioId);
-      if (scenarioFromUrl) handleScenarioChange(scenarioFromUrl.id);
+    if (
+      currentScenarioId &&
+      previousCurrentScenarioId.current !== currentScenarioId &&
+      currentScenarioId !== routerScenarioId
+    ) {
+      navigate(currentScenarioId);
     }
```
Comment (Guismus) [inline src/hooks/RouterHooks.js:88]: useless brackets on if statement

File: src/hooks/RouterHooks.js [hunk 26]
```diff
@@ -29,47 +30,96 @@ export const useRedirectFromInstanceToSenarioView = () => {
...
+  const blocker = useBlocker(when);
+
+  const callOnRouteChange = useCallback(async () => {
+    const dialogResult = await TwoActionsDialogService.openDialog(dialogProps);
+    const proceed = dialogResult === 2;
+    if (proceed) {
+      blocker.proceed();
+    } else {
+      blocker.reset();
+    }
```
Comment (Guismus) [inline src/hooks/RouterHooks.js:103]: useless brackets
`if (proceed) blocker.proceed();
 else blocker.reset();`
 seems better

File: src/hooks/RouterHooks.js [hunk 27]
```diff
@@ -29,47 +30,96 @@ export const useRedirectFromInstanceToSenarioView = () => {
...
+  }, [blocker, dialogProps]);
+
+  useEffect(() => {
+    if (blocker.state === 'blocked') {
+      callOnRouteChange();
+    }
```
Comment (Guismus) [inline src/hooks/RouterHooks.js:109]: useless brackets on if statement

File: src/hooks/RouterHooks.spec.js [hunk 28]
```diff
@@ -0,0 +1,59 @@
+/* eslint-disable react/prop-types */
```
Comment (Guismus) [inline src/hooks/RouterHooks.spec.js:1]: missing copyright header

File: cypress/integration/brewery/DiscardAndContinue.spec.js (hunk 20 above)
Comment (ardvi) [reply cypress/integration/brewery/DiscardAndContinue.spec.js:4]: done

File: src/components/CurrentScenarioSelector/CurrentScenarioSelector.js (hunk 22 above)
Comment (ardvi) [reply src/components/CurrentScenarioSelector/CurrentScenarioSelector.js]: in fact, this prop is not used ...
Remove prop, done

File: src/components/CurrentScenarioSelector/CurrentScenarioSelector.js (hunk 23 above)
Comment (ardvi) [reply src/components/CurrentScenarioSelector/CurrentScenarioSelector.js]: done

File: src/hooks/RouterHooks.js (hunk 24 above)
Comment (ardvi) [reply src/hooks/RouterHooks.js:51]: i keep it for readability (line too long)

File: src/hooks/RouterHooks.js [hunk 29]
```diff
@@ -29,47 +30,96 @@ export const useRedirectFromInstanceToSenarioView = () => {
...
+  }, [currentScenarioId, handleScenarioChange, navigate, routerScenarioId, sortedScenarioList]);
+
+  // on currentScenario change
+  // > on load, if default view, add scenarioId to route
+  // > after load, navigate to new scenario.
+  const previousCurrentScenarioId = useRef(currentScenarioId);
+  useEffect(() => {
+    if (!routerScenarioId && currentScenarioId) {
+      navigate(currentScenarioId, { replace: true });
+    }
```
[the remark this answers is not in the document]
Comment (ardvi) [reply src/hooks/RouterHooks.js:80]: i keep it for readability

File: src/hooks/RouterHooks.js (hunk 25 above)
Comment (ardvi) [reply src/hooks/RouterHooks.js:88]: i keep it for readability

File: src/hooks/RouterHooks.js (hunk 26 above)
Comment (ardvi) [reply src/hooks/RouterHooks.js]: done

File: src/hooks/RouterHooks.js (hunk 27 above)
Comment (ardvi) [reply src/hooks/RouterHooks.js]: done

File: src/hooks/RouterHooks.spec.js (hunk 28 above)
Comment (ardvi) [reply src/hooks/RouterHooks.spec.js:4]: done

**Review by Guismus — APPROVED** [review]

**Comment by Guismus** [comment]
:+1:

**Review by nborde-CSM — APPROVED** [review]
Two remarks but LGTM :+1

**Pull request merged**